### PR TITLE
feat: add triggers for Convex Frax and Convex alUSD

### PR DESCRIPTION
**Pull request by chad-js**
## Description
Linear: https://linear.app/cozy-finance/issue/PRD-338#comment-778f7478
All Curve contracts: https://curve.fi/contracts, https://curve.readthedocs.io/ref-addresses.html#
FRAX/3CRV: https://etherscan.io/address/0xd632f22692fac7611d2aa1c0d552930d43caed3b
alUSD/3CRV: https://etherscan.io/address/0x43b4FdFD4Ff969587185cDB6f0BD875c5Fc83f8c

### Differences between existing Convex triggers (USDP, tBTC) and these new Convex triggers (FRAX, alUSD)

- For FRAX and alUSD, factory meta pools are used by Curve instead of traditional Curve pools. As a result, we don't need to call `minter()` on the LP token contract to get the meta pool contract, like we do in `Convex.sol` for the existing Convex triggers which use traditional Curve meta pools https://github.com/ScopeLift/cozy-triggers/blob/master/contracts/Convex.sol#L90
   - > Factory pools differ from traditional Curve pools in that the pool contract is also the LP token. This improves gas efficiency and simplifies the factory deployment process. https://curve.readthedocs.io/factory-pools.html?highlight=lp%20token#lp-tokens, https://curve.readthedocs.io/ref-addresses.html#promoted-factory-pools

- For factory meta pools, it seems the base pool cannot be determined programmatically, like it can for the traditional meta pools (like we do for USDP and tBTC [here](https://github.com/ScopeLift/cozy-triggers/blob/master/contracts/Convex.sol#L91)). Instead, the base pool can be found in the contract code (ie for [FRAX](https://etherscan.io/address/0xd632f22692fac7611d2aa1c0d552930d43caed3b#code),  search for `BASE_POOL`).
    - Note: There are only two possible base pools as of time of writing ([ref](https://curve.readthedocs.io/factory-deployer.html?highlight=0xbEbc44782C7dB0a1A60Cb6fe97d0b483032FF1C7#base-pools))

- For the existing USDP and tBTC triggers, we need to invoke the base pool's contract functions `coins` and `balances` with different encoded function selectors, resulting in two contract implementations https://github.com/ScopeLift/cozy-triggers/blob/master/contracts/Convex.sol#L200-L253. For FRAX and alUSD, the base pool is the same ([3pool](https://etherscan.io/address/0xbEbc44782C7dB0a1A60Cb6fe97d0b483032FF1C7)), resulting in only one contract implementation needed and simplification of those calls (no need to encode selectors).

## Testing
- re-used existing `Convex.test.js` suite, with some modifications to the `setupFixture` to support `Convex2` contracts (factory meta pools)
    - tested the skipped test (via `it.skip`), `toggles trigger when base pool's get_virtual_price() reverts`
    - reset function, used to re-fork mainnet between tests is used to ensure storage slot state reset between triggers

- deployed to hardhat
Screen Shot 2022-01-07 at 11 35 07 AM

### CODE REVIEW

File: utils/utils.ts
Comment (lvwrence) [inline utils/utils.ts:279]: where's this set?

File: scripts/storage-slots.sh
Comment (lvwrence) [inline scripts/storage-slots.sh:1]: nice

**Comment by lvwrence** [comment]
just ran tests locally + they pass -- this lgtm, but obv should wait for Matt's review!

File: scripts/create-protection-market-convex.ts
Comment (mds1) [inline scripts/create-protection-market-convex.ts]: Usually for testing locally I temporarily import and call the `fundAccount` helper method in the utils file, that way we can test the deploy from the real deployer private key. This just helps minimize the differences between our test deploy and the real deploy

You can do this with `await fundAccount('0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE', '5', signer.address, hre)` to fund the signer with 5 ETH

Also I don't think this line would work anyway since it's not in an async function?

File: scripts/storage-slots.sh
Comment (mds1) [inline scripts/storage-slots.sh]: Just curious why all the code is commented out? I think it'd be good to commit a working version that you can run without modifications

File: test/Convex.test.ts
Comment (mds1) [inline test/Convex.test.ts]: Definitely a nit, but typically lowercase hex is used. It really doesn't matter and isn't specified in the JSON RPC [spec](https://eth.wiki/json-rpc/API), though in examples there you'll see lowercase is used
```suggestion
  '0xd632f22692FaC7611d2AA1C0D552930D43CAEd3B': '0xf', // Curve.fi FRAX/3Crv
  '0x43b4FdFD4Ff969587185cDB6f0BD875c5Fc83f8c': '0xf', // Curve.fi alUSD/3Crv
```

File: test/Convex.test.ts
Comment (mds1) [inline test/Convex.test.ts:191]: Nice job getting it all in one fixture 👌

File: utils/utils.ts
Comment (mds1) [reply utils/utils.ts:279]: it's imported from `hardhat.config.ts`, using the fork URL specified there

**Review by mds1 — COMMENTED** [review]
Looking great!! Tests all pass and deploys are working. Left a few small comments but nothing major

File: scripts/create-protection-market-convex.ts
Comment (chad-js) [reply scripts/create-protection-market-convex.ts]: ahhh, I see. I was a bit stuck with funds with my real private key at first, but I see that `fundAccount` helper would resolve that. 

& yeah lol I had actually executed this within the `main` method. Removed.

File: scripts/storage-slots.sh
Comment (chad-js) [reply scripts/storage-slots.sh]: I commented it out because I figured the user could determine if they wanted either/both balanceOf or totalSupply storage slots and uncomment. But yeah agree, status quo should be uncommented and working seemlessly

File: test/Convex.test.ts
Comment (chad-js) [reply test/Convex.test.ts]: updated

**Review by mds1 — APPROVED** [review]
Everything LGTM! Approving this, but will hold off until merging until after we deploy

**Comment by mds1** [comment]
Merging, and will track deployment separately

**Pull request merged**
